File: DesignPatterns/Typescript/Creators.ts
/*******************\
 * Creators clases *
\*******************/

import {Product, Student, Admin} from "./Products"


// Interfaces
// ------------------------------------------------------------

export interface configObj {
  name: string;
  enrollment: string;
  career?: string;
  grade?: number;
  group?: string;
  job?: string;
  area?: string;
}

// Classes
// ------------------------------------------------------------
export abstract class Creator {

  public abstract factoryMethod(config: configObj): Product;

  public someOperation(): string {
    return 'This a creation!'
  }

} 

export class StudentsCreator extends Creator {
  
  factoryMethod(config: configObj) {
    const { name, enrollment, career, grade, group } = config;
    return new Student(name, enrollment, career, grade, group);
  }

}

export class AdminCreator extends Creator {
  
  factoryMethod(config:configObj) {
    const { name, enrollment, job, area } = config;
    return new Admin(name, enrollment, area, job);
  }

}

File: DesignPatterns/FactoryMethod/Python/Creators.py
""" Creator class """

from __future__ import annotations
from abc import ABC, abstractmethod
from Products import Admin, Product, Student


class Creator(ABC):

    @abstractmethod
    def factory_method(self) -> Product:
        pass

    def some_operation(self) -> str:
        result = "This is some operations"
        return result


class StudentCreator(Creator):

    def factory_method(self, name: str, enrollment: str,
                       career: str, grade: int, group: str) -> Product:
        return Student(name, enrollment, career, grade, group)


class AdminCreator(Creator):

    def factory_method(self, name: str, enrollment: str,
                       area: str, job: str) -> Product:
        return Admin(name, enrollment, area, job)
